Mathlib/Tactic/Linarith/Preprocessing.lean

```diff
@@ -340,15 +345,16 @@ This preprocessor is typically run last, after all inputs have been canonized.
 def nlinarithExtras : GlobalPreprocessor where
   name := "nonlinear arithmetic extras"
   transform ls := do
-    let s ← ls.foldrM (fun h s' => do findSquares s' (← instantiateMVars (← inferType h)))
-      HashSet.empty
-    let new_es ← s.foldM (fun new_es (⟨e, is_sq⟩ : Expr × Bool) =>
-      ((do
-        let p ← mkAppM (if is_sq then ``sq_nonneg else ``mul_self_nonneg) #[e]
-        pure <| p::new_es) <|> pure new_es)) ([] : List Expr)
+    -- find the squares in `AtomM` to ensure deterministic behavior
+    let s ← AtomM.run .reducible do
+      let si ← ls.foldrM (fun h s' => do findSquares s' (← instantiateMVars (← inferType h)))
+        HashSet.empty
+      si.toList.mapM fun (i, is_sq) => return ((← get).atoms[i]!, is_sq)
```

Comment:
To make this more deterministic, could we sort `si.toList`? Otherwise it comes out in whatever order the hash turns out to be. Let's sort by the lex order of the pairs.

Reply:
How about using an `RBSet` instead, so that we can do the sorting as we go?

Reply:
Sure, seems reasonable.